### PR TITLE
feat(webui): add column visibility persistence

### LINKED ISSUE #3570: Column selections not saving in Web View
# Issue
When rerunning evals in promptfoo, the webview automatically updates but hidden column settings are not retained between runs. So when using a large number of variables, I have to re-hide columns that I don't want to display every time I run a new eval. This requires a lot of clicking each time. This was mentioned in below thread, but can't find any other related threads of this issue being reported. 

# Related Background

              @DarioBernardo unfortunately this PR was closed without merging it. Hiding columns and remembering state is on our backlog. We will get to it soon.

_Originally posted by @mldangelo in https://github.com/promptfoo/promptfoo/issues/1069#issuecomment-2383552426_

### PR DESCRIPTION
**Pull request by mldangelo**
## Summary

Adds config-driven default column visibility to the current Web UI's already persisted column selection flow.

- Adds the optional `defaultColumnVisibility` configuration, public schema, documentation, and example.
- Applies configuration defaults only when a saved browser preference does not already exist for the variable schema.
- Supports group defaults plus named overrides, with `showColumns` taking precedence over `hideColumns`.

## Design Decisions

1. Saved browser choices remain authoritative after a user customizes the table. Configuration acts as the initial/default policy rather than overwriting a local preference.
2. Partial configuration is valid. Omitted `variables` or `prompts` settings remain visible, matching viewer behavior and the published JSON schema.
3. Configuration recovered from persisted/API data is defensively parsed before being used by the UI.
4. Named visibility sets are allocated once per resolution pass rather than once per rendered column.

## Test Plan

- [x] Merged current `main`.
- [x] Regenerated `site/static/config-schema.json`.
- [x] `npm run f` and `npm run l` (the existing config-loader complexity warnings remain non-blocking).
- [x] `npm run tsc`.
- [x] `npm run test:app -- src/pages/eval/components/columnVisibility.test.ts src/pages/eval/components/ResultsView.test.tsx --run` (`56` tests passed).
- [x] `npx vitest test/types/index.test.ts test/config-schema.test.ts test/util/config/load.test.ts --run` (`487` tests passed).
- [x] `npm run build`.
- [x] `SKIP_OG_GENERATION=true npm --prefix site run build`.
- [x] `npm run test:app -- --run` (`4,554` tests passed).
- [x] `npm test -- --run` (`18,192` tests passed, `10` skipped).
- [x] Ran an isolated local CLI evaluation using a partial column-visibility configuration; it passed and persisted only disposable state/database files with mode `600`.

No database migrations are introduced by this PR.

Closes #3570

### CODE REVIEW

**Comment by nidhishgajjar** [comment]
**Orb Code Review** (powered by GLM-4.7 on [Orb Cloud](https://orbcloud.dev))

## Summary

This PR fixes column visibility feedback issues in the web UI. The changes include fixing `handleShowAll` to clear name-based preferences before showing all columns, fixing `handleToggleVariables` to only update global defaults (not individual preferences), renaming `updateColumnVisibility` to `handleColumnSelectorChange` with clearer documentation, and adding comprehensive tests for the ColumnSelector component.

## Architecture

**The fixes:**

1. **Fixed `handleShowAll`** (`ColumnSelector.tsx`):
```typescript
// Before: Show all columns without clearing preferences
const handleShowAll = () => {
  const allColumns = columnData.map((col) => col.value);
  onChange(allColumns);
};

// After: Clear name-based preferences first, then show all columns
const handleShowAll = () => {
  onResetAllPreferences();  // Clear name-based preferences
  onSetGlobalVariableVisibility(true);  // Show all variables
  onSetGlobalPromptVisibility(true);  // Show all prompts
  const allColumns = columnData.map((col) => col.value);
  onChange(allColumns);
};
```

2. **Fixed `handleToggleVariables`**:
```typescript
// Before: Updated both global and individual preferences
const handleToggleVariables = () => {
  const newGlobalVisibility = !globalVariableVisibility;
  onSetGlobalVariableVisibility(newGlobalVisibility);
  // Also updated individual variable preferences
};

// After: Only update global default
const handleToggleVariables = () => {
  const newGlobalVisibility = !globalVariableVisibility;
  onSetGlobalVariableVisibility(newGlobalVisibility);
  // Individual variable preferences are not modified
};
```

3. **Renamed and documented**:
```typescript
// Before: Unclear name and purpose
const updateColumnVisibility = (column) => {
  // ...
};

// After: Clear name and documentation
const handleColumnSelectorChange = (column) => {
  // Handle checkbox change for individual column
  // Updates individual column preference via onSaveColumnPreference
  // If column is being unchecked and it's the last one checked,
  //   prevent unchecking to ensure at least one column remains visible
  if (!column.checked && selectedColumns.length === 1) {
    return;  // Prevent unchecking last column
  }
  onSaveColumnPreference(column.value, column.checked);
};
```

4. **Added comprehensive tests** (`ColumnSelector.test.tsx`):
- Tests for `handleShowAll`: clears preferences, calls global setters, calls onChange with all columns
- Tests for `handleToggleVariables`: only calls global setter, not individual preferences
- Tests for column selector interaction: prevents unchecking last column, handles preferences correctly
- Total: 268 lines of new tests

**Key improvements:**

1. **Better UX**: `handleShowAll` now properly resets preferences before showing all columns

2. **Correct semantics**: `handleToggleVariables` only affects global defaults, not individual prefs

3. **Clearer code**: Renamed function and added comprehensive documentation

4. **Comprehensive tests**: 268 lines of tests ensure the fixes work correctly

5. **Prevents edge cases**: Prevents unchecking the last column (ensures at least one column visible)

## Issues

**No issues found.**

This is a well-implemented fix that:

1. **Solves real UX issues**: Fixes column visibility behavior based on user feedback

2. **Correct semantics**: Separates global defaults from individual preferences properly

3. **Clear code**: Renamed function and added good documentation

4. **Comprehensive tests**: 268 lines of tests cover all the fixed behaviors

5. **Prevents edge cases**: Last column protection ensures good UX

## Cross-file Impact

**Affected components:**
- `src/app/src/pages/eval/components/ColumnSelector.tsx` — Fixed column visibility logic (21 insertions, 1 deletion)
- `src/app/src/pages/eval/components/ResultsView.tsx` — Updated to use new function name (36 insertions, 28 deletions)
- `src/app/src/pages/eval/components/ColumnSelector.test.tsx` — NEW: Comprehensive tests (268 insertions)

**Breaking changes:** None. This is a UX improvement.

**API changes:**
- Renamed: `updateColumnVisibility` → `handleColumnSelectorChange`
- New parameters: `onResetAllPreferences`, `onSetGlobalVariableVisibility`, `onSetGlobalPromptVisibility`

**Behavior changes:**
- **Before**: `handleShowAll` showed all columns without clearing preferences
- **After**: `handleShowAll` clears name-based preferences first, then shows all columns
- **Before**: `handleToggleVariables` updated both global and individual preferences
- **After**: `handleToggleVariables` only updates global defaults
- **Before**: Individual column changes were unclear
- **After**: Individual column changes have clear semantics and prevent edge cases

**UX improvements:**
- Better column visibility management
- Clearer separation between global defaults and individual preferences
- Prevents accidentally hiding all columns

## Assessment

✅ Approve

Well-implemented UX improvement with comprehensive testing:

**Strengths:**
- Solves real user feedback about column visibility issues
- Correct semantics for global vs individual preferences
- Clear function naming and documentation
- Comprehensive test coverage (268 lines)
- Prevents edge cases (last column protection)
- Properly updates calling code to use new function name

**User benefit:**
Users now have better control over column visibility with clear semantics:
- "Show All" properly resets preferences and shows all columns
- Variable/Prompt toggles only affect global defaults
- Individual column preferences are properly managed
- Cannot accidentally hide all columns

The implementation is correct, well-tested, and addresses the stated UX issues.

**Comment by mldangelo-oai** [comment]
@codex review

File: src/app/src/pages/eval/components/columnVisibility.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    allColumns.push('description');
+    // Description follows variable visibility rules
+    visibility['description'] =
+      namePrefs['description'] ??
+      (configHiddenColumns.has('description')
+        ? false
+        : configShownColumns.has('description')
+          ? true
+          : configShowVars && defaults.showAllVariables);
+  }
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply src/app/src/pages/eval/components/columnVisibility.ts]: Resolved on the current-main merged tip: visibility now flows through the shared resolver, where showColumns wins for every column type; added explicit description and prompt precedence coverage.

File: src/app/src/pages/eval/components/columnVisibility.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+  return { columnVisibility, selectedColumns };
+}
+
+export function getConfigColumnVisibility(
+  config?: Partial<UnifiedConfig> | null,
+): DefaultColumnVisibility | undefined {
+  return config?.defaultColumnVisibility as DefaultColumnVisibility | undefined;
+}
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply src/app/src/pages/eval/components/columnVisibility.ts:102]: Resolved on the pushed tip: persisted/API defaults are parsed with DefaultColumnVisibilitySchema.safeParse before use, with malformed-value coverage.

File: src/types/index.ts [hunk 3]
```diff
@@ -1193,6 +1193,21 @@ export const TestSuiteConfigSchema = z.object({
...
+  // Controls which columns are shown/hidden by default when viewing eval results
+  defaultColumnVisibility: z
+    .object({
+      // Whether to show all variables by default ('visible') or hide them ('hidden')
+      variables: z.enum(['visible', 'hidden']).default('visible'),
+      // Whether to show all prompts by default ('visible') or hide them ('hidden')
+      prompts: z.enum(['visible', 'hidden']).default('visible'),
+      // Specific variable columns to hide (by variable name, e.g., 'context', 'system_prompt')
+      hideColumns: z.array(z.string()).optional(),
+      // Specific variable columns to show (only applies when variables: 'hidden')
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply src/types/index.ts]: Resolved on the pushed tip: the duplicate misleading inline schema comment is removed; docs now state showColumns overrides hideColumns for named columns.

File: examples/column-visibility/promptfooconfig.yaml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+providers:
+  - openai:gpt-4o-mini
+
+# Hide verbose columns by default in the web viewer
+defaultColumnVisibility:
+  variables: visible
+  hideColumns:
+    - context
+    - system_prompt
+
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply examples/column-visibility/promptfooconfig.yaml:22]: Resolved on the pushed tip: the example includes prompts: visible, and the regenerated public schema now correctly accepts partial defaults matching runtime fallback behavior.

File: src/app/src/pages/eval/components/columnVisibility.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  }
+
+  const shownColumns = new Set(configDefaults.showColumns ?? []);
+  if (isNamed(shownColumns, semanticName, columnId)) {
+    return true;
+  }
+
+  const hiddenColumns = new Set(configDefaults.hideColumns ?? []);
+  if (isNamed(hiddenColumns, semanticName, columnId)) {
+    return false;
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply src/app/src/pages/eval/components/columnVisibility.ts]: Resolved on the pushed tip: show/hide sets are allocated once per visibility resolution rather than once per column.

File: src/types/index.ts [hunk 6]
```diff
@@ -1193,6 +1193,21 @@ export const TestSuiteConfigSchema = z.object({
...
+      showColumns: z.array(z.string()).optional(),
+    })
+    .optional(),
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply src/types/index.ts]: Resolved on the pushed tip: DefaultColumnVisibilitySchema is defined once and reused by TestSuiteConfigSchema and runtime extraction.

**Comment by mldangelo-oai** [comment]
Merged current `main` and pushed `080884653350d4e39cfd631caf4948fa40f0121e`.

What changed:
- Resolved the current-main conflict in `test/util/config/load.test.ts` by preserving both the provider-dedupe regression coverage from `main` and the `defaultColumnVisibility` merge coverage from this PR.
- Clarified the docs/example wording so `hideColumns`/`showColumns` distinguish semantic variable names from display column IDs, and so browser preference scope matches the implementation.
- Regenerated the config schema against the merged tree; no additional schema diff was needed.

Review/comments:
- No unresolved, non-outdated review threads remain on the current PR data.
- No database migrations are introduced.

Validation run:
- `npm run jsonSchema:generate`
- `npm run test:app -- src/pages/eval/components/columnVisibility.test.ts --run` (14 passed)
- `npm run test:app -- src/pages/eval/components/ResultsView.test.tsx --run` (42 passed)
- `npx vitest --run test/types/index.test.ts test/config-schema.test.ts test/index.test.ts test/util/config/load.test.ts` (552 passed)
- `npm run f`
- `npm run l` (passed; existing config-loader complexity warnings only)
- `npm run tsc`
- `npm run build`
- `SKIP_OG_GENERATION=true npm --prefix site run build`
- Isolated CLI E2E with `PROMPTFOO_CONFIG_DIR=/tmp/pr6874-state` and `PROMPTFOO_CACHE_PATH=/tmp/pr6874-cache`: local echo eval passed 1/1 and exported `/tmp/pr6874-e2e/output.json` with `config.defaultColumnVisibility` preserved (`variables: visible`, `prompts: hidden`, `hideColumns: ["context"]`, `showColumns: ["question"]`).

Fresh CI is queued/running on the new head; moving to the next PR per the loop instruction.

File: src/types/index.ts [hunk 7]
```diff
@@ -1227,6 +1237,10 @@ export const TestSuiteConfigSchema = z.object({
   // Write results to disk so they can be viewed in web viewer
   writeLatestResults: z.boolean().optional(),
 
+  // Default column visibility settings for the web viewer
+  // Controls which columns are shown/hidden by default when viewing eval results
+  defaultColumnVisibility: DefaultColumnVisibilitySchema.optional(),
```
Comment (jameshiester-oai) [inline src/types/index.ts:1242]: **[P2] Propagate the new field through Web UI eval jobs**

This makes `defaultColumnVisibility` part of the public config, but the Web UI serializers do not carry it: `RunTestSuiteButton.tsx:25-36` omits it from the extracted config and lines 75-86 omit it from the `/eval/job` body; `stores/evalConfig.ts:45-61` omits it from `getTestSuite()` as well. Loading a config with this field into setup and running/editing it therefore silently removes the defaults from the saved eval. Add the field to the shared job/YAML serialization path and assert it in the request-boundary tests.

File: src/app/src/pages/eval/components/columnVisibility.ts [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  const match = columnId.match(/^Variable (\d+)$/);
+  if (!match) {
+    return null;
+  }
+  const varIndex = parseInt(match[1], 10) - 1;
+  return varNames[varIndex] ?? null;
+}
+
+function isNamed(set: Set<string>, semanticName: string, columnId: string): boolean {
+  return set.has(semanticName) || set.has(columnId);
```
Comment (jameshiester-oai) [inline src/app/src/pages/eval/components/columnVisibility.ts:32]: **[P2] Give semantic names and column IDs distinct namespaces**

Variable names are arbitrary, so matching every selector against both values makes valid names collide with unrelated display IDs. For `varNames: ["description"]`, `allColumns: ["description", "Variable 1", "Prompt 1"]`, and `hideColumns: ["description"]`, both the Description column and Variable 1 are hidden; a variable named `Prompt 1` similarly hides both that variable and Prompt 1. The documented syntax cannot target only one. Introduce an unambiguous/namespaced selector form (while preserving non-colliding legacy values) and add reserved-name collision tests.

**Review by jameshiester-oai — COMMENTED** [review]
Reviewed base `87d04046debc840006659f39b1e6481479d260ee` at head `080884653350d4e39cfd631caf4948fa40f0121e`.

**Scope:** fresh full review; no authenticated completion marker existed for this head. I reviewed the full base...head diff, surrounding Web UI config/persistence/job code, schemas, generated schema, docs/examples, tests, all prior threads, and exact-head CI.

**Verdict:** the CLI/config pipeline works, but two user-visible Web UI defects remain. Exact-head CI is green: 41/41 checks completed successfully.

## Verified findings

### P2

- The Web UI create/edit-and-rerun path drops `defaultColumnVisibility`. `RunTestSuiteButton.tsx:25-36` does not extract it and the `/eval/job` object at lines 75-86 does not send it; `stores/evalConfig.ts:45-61` also omits it from `getTestSuite()`, which feeds the YAML preview. A direct store probe retained the field in `config` but returned `undefined` from `getTestSuite()`. As a result, configs loaded into the setup UI silently lose these defaults when previewed or run, and the resulting eval cannot apply the feature. See the inline comment on the new schema field.
- Selector strings share one namespace for semantic variable names and synthetic column IDs. Valid variables named `description`, `Prompt 1`, or `Variable 2` therefore affect both the variable and unrelated display column, with no syntax to target one independently. See the inline comment on `columnVisibility.ts`.

## Existing feedback

- All prior inline threads are resolved. I rechecked the current implementations for description precedence, partial schema defaults, runtime parsing, set allocation, example validity, and multi-config merge order; none of those findings remains current.

## Tests and QA

- `npm run test:app -- src/pages/eval/components/columnVisibility.test.ts src/pages/eval/components/ResultsView.test.tsx --run` — 2 files, 56 tests passed.
- `npx vitest run test/types/index.test.ts test/config-schema.test.ts test/index.test.ts test/util/config/load.test.ts --sequence.shuffle=false` — 4 files, 552 tests passed.
- `npm run test:app -- src/pages/eval-creator/components/RunTestSuiteButton.test.tsx src/stores/evalConfig.test.ts --run` — 2 files, 25 tests passed; these suites currently do not assert the new field.
- `npm run build` — passed, including TypeScript and the app production build.
- `SKIP_OG_GENERATION=true npm --prefix site run build` — passed.
- `git diff --check 87d04046debc840006659f39b1e6481479d260ee...HEAD` — passed.
- `PROMPTFOO_DISABLE_REMOTE_GENERATION=true npm run local -- eval -c /tmp/pr6874-review/promptfooconfig.yaml --no-cache -o /tmp/pr6874-review/results.json` — exit 0; 1 passed, 0 failed, 0 errors; export preserved the partial `defaultColumnVisibility` object.
- Direct exact-head probes reproduced selector collisions for `description`, `Prompt 1`, and `Variable 2`, and reproduced the Web UI store serializer dropping the field.

## Simplification and coverage

- Use one typed Web UI-to-job serializer instead of maintaining separate field lists in `getTestSuite()` and `RunTestSuiteButton`; this removes the drift that caused the first defect.
- Add request-boundary coverage asserting `defaultColumnVisibility` in `/eval/job` payloads and collision cases for reserved-looking variable names.

## Limitations

- Browser QA was not run: port 3000 is an SSH listener and no Promptfoo server is listening on 15500; repository guidance requires confirmation before starting the development server. Pure resolver tests, component tests, request/store tests, production builds, and CLI pipeline QA were run instead.
- The `/pr-review-toolkit:review-pr` command contract was present locally but not exposed as a callable Codex skill endpoint. Its correctness, tests, comments/docs, error-handling, type-design, and simplification lenses were executed through the orchestrated review and three independent read-only specialist passes.
